src/util/bson/bson.cpp

```diff
@@ -443,641 +444,196 @@ std::ostream& operator<<(std::ostream& out, const Bson& b)
- @param[in] val  object key
- @return whether parsing should proceed
- @note It is safe to move the passed string.
- */
-bool Parser::key(string_t& val) {
-    if (!m_instructions.empty()) {
-        auto top = m_instructions.top();
-
-        if (top.type == State::RegularExpression) {
-            if (val == key_regular_expression_pattern) {
-                m_instructions.push({State::RegularExpressionPattern, top.key});
-            } else if (val == key_regular_expression_options) {
-                m_instructions.push({State::RegularExpressionOptions, top.key});
-            }
-            return true;
-        } else if (top.type == State::Date) {
-            return true;
-        } else if (top.type == State::Binary) {
-            if (val == key_binary_base64) {
-                m_instructions.push({State::BinaryBase64, top.key});
-            } else if (val == key_binary_sub_type) {
-                m_instructions.push({State::BinarySubType, top.key});
+    REALM_TERMINATE("unknown json value type");
+}
+
+// This works around the deleted rvalue constructor in StringData
+inline StringData tosd(const std::string& s) { return s; }
+
+// Keep these sorted by key. This is checked so you can't forget.
+using FancyParser = Bson(*)(const Json& json);
+static constexpr std::pair<std::string_view, FancyParser> bson_fancy_parsers[] = {
+    {"$binary", +[](const Json& json) {
+        util::Optional<std::vector<char>> base64;
+        util::Optional<uint8_t> subType;
+        if (json.size() != 2)
+            throw BsonError("invalid extended json $binary");
+        for (auto&& [k, v] : json.items()) {
+            if (k == "base64") {
+                const std::string& str = v.get<std::string>();
+                base64.emplace(str.begin(), str.end());
+            } else if (k == "subType") {
+                subType = std::stoul(v.get<std::string>(), nullptr, 16);
             }
-            return true;
-        } else if (top.type == State::Timestamp) {
-            if (val == key_timestamp_time) {
-                m_instructions.push({State::TimestampT, top.key});
-            } else if (val == key_timestamp_increment) {
-                m_instructions.push({State::TimestampI, top.key});
-            }
-            return true;
         }
-    }
-
-    const auto it = bson_type_for_key.find(val.data());
-    const auto type = (it != bson_type_for_key.end()) ? (*it).second : Parser::State::JsonKey;
-
-    // if the key denotes a bson type
-    if (type != State::JsonKey) {
-        m_marks.pop();
-
-        if (m_instructions.size()) {
-            // if the previous instruction is a key, we don't want it
-            if (m_instructions.top().type == State::JsonKey) {
-                m_instructions.pop();
+        if (!base64 || !subType)
+            throw BsonError("invalid extended json $binary");
+        return Bson(std::move(*base64)); // TODO don't throw away the subType.
+    }},
+    {"$date", +[](const Json& json) {
```

Comment:
What does the `+` signify here?

Reply:
It decays the lambda object (each lambda expression creates a callable object of a unique type) to an equivalent plain function pointer. This is only possible when the lambda has no captures (ie when it is just fancy syntax for a plain function, not a closure). It probably isn't actually necessary here because the conversion should happen implicitly as part of initializing the array. It is mostly habit to put it there when I want it to happen. It is more important when you are passing a lambda to a context where the type is deduced like `auto f = +[]{};` or `template <typename T> f(T&&); f(+[]{});` or even `std::function(+[]{})`.

Reply:
Ah, that makes sense.